src/net/server/middleware/tsig.rs

```diff
@@ -0,0 +1,496 @@
+where
+    KS: Clone + KeyStore,
+{
+    /// Creates an instance of this middleware service.
+    ///
+    /// Keys in the provided [`KeyStore`] will be used to verify received signed
+    /// requests and to sign the corresponding responses.
+    #[must_use]
+    pub fn new(next_svc: NextSvc, key_store: KS) -> Self {
+        Self {
+            next_svc,
+            key_store,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<RequestOctets, NextSvc, KS> TsigMiddlewareSvc<RequestOctets, NextSvc, KS>
+where
+    RequestOctets: Octets + OctetsFrom<Vec<u8>> + Send + Sync + Unpin,
+    NextSvc: Service<RequestOctets, Option<KS::Key>>,
+    NextSvc::Target: Composer + Default,
+    KS: Clone + KeyStore,
+    KS::Key: Clone,
+    Infallible: From<<RequestOctets as octseq::OctetsFrom<Vec<u8>>>::Error>,
+{
+    #[allow(clippy::type_complexity)]
+    fn preprocess(
+        req: &Request<RequestOctets>,
+        key_store: &KS,
+    ) -> ControlFlow<
+        AdditionalBuilder<StreamTarget<NextSvc::Target>>,
+        Option<(
+            Request<RequestOctets, Option<KS::Key>>,
+            TsigSigner<KS::Key>,
+        )>,
+    > {
+        if let Some(q) = Self::get_relevant_question(req.message()) {
+            let octets = req.message().as_slice().to_vec();
+            let mut mut_msg = Message::from_octets(octets).unwrap();
+
+            match tsig::ServerTransaction::request(
+                key_store,
+                &mut mut_msg,
+                Time48::now(),
+            ) {
+                Ok(None) => {
+                    // Message is not TSIG signed.
+                }
+
+                Ok(Some(tsig)) => {
+                    // Message is TSIG signed by a known key.
+                    trace!(
+                        "Request is signed with TSIG key '{}'",
+                        tsig.key().name()
+                    );
+
+                    // Convert to RequestOctets so that the non-TSIG signed
+                    // message case can just pass through the RequestOctets.
+                    let source = mut_msg.into_octets();
+                    let octets = RequestOctets::octets_from(source);
+                    let new_msg = Message::from_octets(octets).unwrap();
+
+                    let mut new_req = Request::new(
+                        req.client_addr(),
+                        req.received_at(),
+                        new_msg,
+                        req.transport_ctx().clone(),
+                        Some(tsig.key_wrapper().clone()),
+                    );
+
+                    let num_bytes_to_reserve = tsig.key().compose_len();
+                    new_req.reserve_bytes(num_bytes_to_reserve);
+
+                    return ControlFlow::Continue(Some((
+                        new_req,
+                        TsigSigner::Transaction(tsig),
+                    )));
+                }
+
+                Err(err) => {
```

Comment:
This is wrong. `build_message()` should be called on the received error.

Reply:
I pushed a fix for this. @tertsdiepraam, as you worked in this area recently on PR #334 you might also want to take a look.

Reply:
Hmm. I couldn't find the fix I thought I pushed for this so I just pushed it now.

Reply:
@tertsdiepraam: Can I resolve this conversation?